apps/snitch_core/lib/core/domain/shipment_engine.ex

```diff
@@ -0,0 +1,146 @@
+defmodule Snitch.Domain.ShipmentEngine do
+  @moduledoc """
+  Finds the optimal shipment for a given order.
+
+  `ShipmentEngine` models the problem as a [Constraint Satisfaction
+  Problem][csp] and find the optimal `shipment` subject to the following
+  constraints:
+
+  1. Two (or more) `packages` cannot fulfill the same `lineitem`.
+  2. Selected `packages` together fulfill the entire `order`.
+
+  Returns a shipment, which is a list of `packages`.
+  In case we are unable to find such a shipment, the empty `list` is returned.
+
+  ## Limitations and future work
+
+  Constraints that could be added in future:
+  * prefer packages that are "on-hand" over backorder packages.
+  * prefer shorter shipping distance (analogous to shipping time)
+  * prefer smaller shipping cost (to user/store owner)
+
+  [csp]: https://en.wikipedia.org/wiki/Constraint_Satisfaction_Problem
+  """
+
+  alias Aruspex.Problem
+  alias Snitch.Data.Schema.Order
+  alias Aruspex.Strategy.SimulatedAnnealing
+
+  @domain [true, false]
+
+  @doc """
+  Returns the optimal shipment from the `packages` fulfilling the `order`.
+
+  The CSP is modelled as a graph of packages, where 2 packages are linked by an
+  edge if they include the same line-item.
+
+  Uses [Simulated Annealing][sa] to find the optimal shipment, see
+  `Aruspex.Strategy.SimulatedAnnealing`. Note that this technique is
+  probabilistic and [AC-3][ac3] should replace it.
+
+  [sa]: https://en.wikipedia.org/wiki/Simulated_annealing
+  [ac3]: https://en.wikipedia.org/wiki/AC-3_algorithm
+  """
+  @spec run(list, Order.t()) :: list
+  def run(packages, %Order{} = order) do
+    packages_with_id = append_keys(packages)
```

Comment:
`append_keys` is called in `create_csp`.
Is there a need to call it again?

Reply:
I need the packages with id to create a list of vars for the CSP.